packages/gatsby/src/redux/nodes.ts

```diff
@@ -688,29 +698,32 @@ export const getNodesFromCacheByValue = (
     }
     const filterValueArr: Array<FilterValueNullable> = filterValue
 
-    const set = new Set<IGatsbyNode>()
+    const arr: Array<IGatsbyNode> = []
+
+    // Note: it's very unlikely that the list of filter values is big so .includes should be fine here
     if (filterValueArr.includes(null)) {
       // Like all other ops, `in: [null]` behaves weirdly, allowing all nodes
       // that do not actually have a (complete) path (v=undefined)
       const nodes = filterCache.byValue.get(undefined)
       if (nodes) {
-        nodes.forEach(v => set.add(v))
+        nodes.forEach(v => arr.push(v))
       }
     }
 
     // For every value in the needle array, find the bucket of nodes for
-    // that value, add this bucket of nodes to one set, return the set.
+    // that value, add this bucket of nodes to one list, return the list.
     filterValueArr
       .slice(0) // Sort is inline so slice the original array
       .sort((a, b) => {
         if (a == null || b == null) return 0
         return a < b ? -1 : a > b ? 1 : 0
       }) // Just sort to preserve legacy order as much as possible.
       .forEach((v: FilterValueNullable) =>
-        filterCache.byValue.get(v)?.forEach(v => set.add(v))
+        filterCache.byValue.get(v)?.forEach(v => arr.push(v))
       )
 
-    return set
+    // TODO: proper merge sort
+    return arr.sort((A, B) => (A.id < B.id ? -1 : A.id === B.id ? 0 : 1))
```

Comment:
Do we still need to sort `filterValueArr` above if we re-sort the resulting `arr` here?

Reply:
No, good point :)